File: packages/android/storage-android/src/test/kotlin/app/notes/storage/GoogleDriveAdapterTest.kt
package app.notes.storage

import app.notes.domain.FileInput
import app.notes.domain.SIMPLE_UPLOAD_THRESHOLD_BYTES
import app.notes.domain.StorageError
import app.notes.domain.StorageException
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import java.util.concurrent.atomic.AtomicInteger

private val ZERO_BACKOFF = BackoffConfig(
    maxAttempts = 3,
    baseDelayMs = 0,
    maxDelayMs = 0,
    delay = { _ -> },
    random = { 0.0 },
)

class GoogleDriveAdapterTest {
    private val tokenProvider: TokenProvider = { "ya29.token" }

    @Test
    fun `ensureAppFolder creates the folder when none exists and caches the id`() = runTest {
        val http = FakeDriveHttp(findResult = null)
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        val id1 = adapter.ensureAppFolder()
        val id2 = adapter.ensureAppFolder()
        assertEquals(id1, id2)
        assertEquals("folder-xyz", id1)
        assertEquals(1, http.findCalls)
        assertEquals(1, http.createCalls)
    }

    @Test
    fun `ensureAppFolder returns the pre-existing folder without creating a duplicate`() = runTest {
        val http = FakeDriveHttp(findResult = "folder-existing")
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        assertEquals("folder-existing", adapter.ensureAppFolder())
        assertEquals(0, http.createCalls)
    }

    @Test
    fun `ensureAppFolder throws Unauthorized when no token is available`() = runTest {
        val adapter = GoogleDriveAdapter(FakeDriveHttp(), { null }, ZERO_BACKOFF)
        val ex = assertThrows(StorageException::class.java) {
            kotlinx.coroutines.runBlocking { adapter.ensureAppFolder() }
        }
        assertTrue(ex.error is StorageError.Unauthorized)
    }

    @Test
    fun `upload uses simpleUpload below the resumable threshold`() = runTest {
        val http = FakeDriveHttp()
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        val remote = adapter.upload(
            FileInput(
                name = "small.pdf",
                mimeType = "application/pdf",
                sizeBytes = 4,
                bytes = byteArrayOf(1, 2, 3, 4),
            ),
        )
        assertEquals("file-1", remote.externalId)
        assertEquals(1, http.simpleUploadCalls)
        assertEquals(0, http.resumableStartCalls)
    }

    @Test
    fun `upload refuses simpleUpload when only a stream is provided under the threshold`() = runTest {
        val http = FakeDriveHttp()
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        val ex = assertThrows(StorageException::class.java) {
            kotlinx.coroutines.runBlocking {
                adapter.upload(
                    FileInput(
                        name = "small.pdf",
                        mimeType = "application/pdf",
                        sizeBytes = 4,
                        stream = { ByteArrayInputStream(byteArrayOf(1, 2, 3, 4)) },
                    ),
                )
            }
        }
        assertTrue(ex.error is StorageError.TransportError)
    }

    @Test
    fun `upload uses resumable path above the threshold and emits final metadata`() = runTest {
        val size = SIMPLE_UPLOAD_THRESHOLD_BYTES + 16
        val bytes = ByteArray(size.toInt())
        val http = FakeDriveHttp(
            putChunkBehaviour = { _, _, rangeStart, total ->
                if (rangeStart + size > total - 1) DriveFileMeta("file-resumable", "big.bin", "application/octet-stream", total)
                else null
            },
        )
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        val remote = adapter.upload(
            FileInput(
                name = "big.bin",
                mimeType = "application/octet-stream",
                sizeBytes = size,
                stream = { ByteArrayInputStream(bytes) },
            ),
        )
        assertEquals(1, http.resumableStartCalls)
        assertTrue(http.putChunkCalls > 0)
        assertEquals(size, remote.sizeBytes)
    }

    @Test
    fun `upload throws TransportError when the stream is shorter than sizeBytes`() = runTest {
        val claimedSize = SIMPLE_UPLOAD_THRESHOLD_BYTES + 1024
        val actualBytes = ByteArray((SIMPLE_UPLOAD_THRESHOLD_BYTES + 16).toInt())
        val http = FakeDriveHttp(putChunkBehaviour = { _, _, _, _ -> null })
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        val ex = assertThrows(StorageException::class.java) {
            kotlinx.coroutines.runBlocking {
                adapter.upload(
                    FileInput(
                        name = "short.bin",
                        mimeType = "application/octet-stream",
                        sizeBytes = claimedSize,
                        stream = { ByteArrayInputStream(actualBytes) },
                    ),
                )
            }
        }
        assertTrue(ex.error is StorageError.TransportError)
    }

    @Test
    fun `getDownloadUrl embeds the token and pre-validates the file exists`() = runTest {
        val http = FakeDriveHttp()
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        val url = adapter.getDownloadUrl("file-abc")
        assertTrue(url.contains("file-abc"))
        assertTrue(url.contains("access_token=ya29.token"))
        assertEquals(1, http.getMetadataCalls)
    }

    @Test
    fun `getDownloadUrl rejects ttl outside the (0, 3600) window`() = runTest {
        val adapter = GoogleDriveAdapter(FakeDriveHttp(), tokenProvider, ZERO_BACKOFF)
        assertThrows(StorageException::class.java) {
            kotlinx.coroutines.runBlocking { adapter.getDownloadUrl("file-abc", ttlSeconds = 0) }
        }
        assertThrows(StorageException::class.java) {
            kotlinx.coroutines.runBlocking { adapter.getDownloadUrl("file-abc", ttlSeconds = 4000) }
        }
    }

    @Test
    fun `retry on 5xx eventually succeeds`() = runTest {
        val attempts = AtomicInteger(0)
        val http = FakeDriveHttp(
            simpleUploadBehaviour = {
                val n = attempts.incrementAndGet()
                if (n < 3) throw StorageException(StorageError.UpstreamError(503, "Service Unavailable"))
                DriveFileMeta("file-retry", "f.bin", "application/octet-stream", 4)
            },
        )
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        val remote = adapter.upload(
            FileInput(name = "f.bin", mimeType = "application/octet-stream", sizeBytes = 4, bytes = ByteArray(4)),
        )
        assertEquals("file-retry", remote.externalId)
        assertEquals(3, attempts.get())
    }

    @Test
    fun `401 is not retried`() = runTest {
        val attempts = AtomicInteger(0)
        val http = FakeDriveHttp(
            simpleUploadBehaviour = {
                attempts.incrementAndGet()
                throw StorageException(StorageError.Unauthorized("Token expired"))
            },
        )
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        assertThrows(StorageException::class.java) {
            kotlinx.coroutines.runBlocking {
                adapter.upload(
                    FileInput("f", "application/octet-stream", 4, byteArrayOf(0, 0, 0, 0)),
                )
            }
        }
        assertEquals(1, attempts.get())
    }

    @Test
    fun `gives up after maxAttempts`() = runTest {
        val attempts = AtomicInteger(0)
        val http = FakeDriveHttp(
            simpleUploadBehaviour = {
                attempts.incrementAndGet()
                throw StorageException(StorageError.UpstreamError(503, "Service Unavailable"))
            },
        )
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        assertThrows(StorageException::class.java) {
            kotlinx.coroutines.runBlocking {
                adapter.upload(
                    FileInput("f", "application/octet-stream", 4, byteArrayOf(0, 0, 0, 0)),
                )
            }
        }
        // initial + 3 retries
        assertEquals(4, attempts.get())
    }

    @Test
    fun `delete forwards the externalId to the wire`() = runTest {
        val http = FakeDriveHttp()
        val adapter = GoogleDriveAdapter(http, tokenProvider, ZERO_BACKOFF)
        adapter.delete("file-abc")
        assertEquals(listOf("file-abc"), http.deletedIds)
    }
}

private class FakeDriveHttp(
    val findResult: String? = null,
    val simpleUploadBehaviour: suspend () -> DriveFileMeta = {
        DriveFileMeta("file-1", "spec.pdf", "application/pdf", 4)
    },
    val putChunkBehaviour: suspend (sessionUri: String, chunk: ByteArray, rangeStart: Long, totalSize: Long) -> DriveFileMeta? = { _, _, _, _ -> null },
) : DriveHttp {
    var findCalls = 0
    var createCalls = 0
    var simpleUploadCalls = 0
    var resumableStartCalls = 0
    var putChunkCalls = 0
    var getMetadataCalls = 0
    val deletedIds = mutableListOf<String>()

    override suspend fun findAppFolder(token: String, parentId: String?, name: String): String? {
        findCalls += 1
        return findResult
    }

    override suspend fun createFolder(token: String, parentId: String?, name: String): String {
        createCalls += 1
        return "folder-xyz"
    }

    override suspend fun simpleUpload(
        token: String,
        folderId: String,
        name: String,
        mimeType: String,
        bytes: ByteArray,
        folderHint: String?,
    ): DriveFileMeta {
        simpleUploadCalls += 1
        return simpleUploadBehaviour()
    }

    override suspend fun startResumableUpload(
        token: String,
        folderId: String,
        name: String,
        mimeType: String,
        sizeBytes: Long,
        folderHint: String?,
    ): ResumableUploadHandle {
        resumableStartCalls += 1
        return ResumableUploadHandle("https://drive/session")
    }

    override suspend fun putChunk(
        sessionUri: String,
        chunk: ByteArray,
        rangeStart: Long,
        totalSize: Long,
    ): DriveFileMeta? {
        putChunkCalls += 1
        return putChunkBehaviour(sessionUri, chunk, rangeStart, totalSize)
    }

    override suspend fun getMetadata(token: String, externalId: String): DriveFileMeta {
        getMetadataCalls += 1
        return DriveFileMeta(externalId, "spec.pdf", "application/pdf", 1024)
    }

    override suspend fun delete(token: String, externalId: String) {
        deletedIds.add(externalId)
    }
}

File: packages/storage/src/__tests__/google-drive-adapter.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { SIMPLE_UPLOAD_THRESHOLD_BYTES, StorageError, StorageException } from "@notes-app/domain";
import { GoogleDriveAdapter } from "../google-drive-adapter.js";
import type { DriveHttpClient, DriveListMatch, ResumableUploadHandle } from "../drive-http.js";

const ZERO_DELAY = {
  maxAttempts: 3,
  baseDelayMs: 0,
  maxDelayMs: 0,
  delay: async () => {},
  random: () => 0,
};

function makeHttp(overrides: Partial<DriveHttpClient> = {}): DriveHttpClient {
  return {
    findAppFolder: vi.fn(async () => null),
    createFolder: vi.fn(async () => "folder-xyz"),
    simpleUpload: vi.fn(async () => sampleMatch("file-1", 4)),
    startResumableUpload: vi.fn(
      async () => ({ sessionUri: "https://drive/session" }) as ResumableUploadHandle,
    ),
    putChunk: vi.fn(async () => null),
    getMetadata: vi.fn(async () => sampleMatch("file-1", 1024)),
    delete: vi.fn(async () => {}),
    ...overrides,
  };
}

function sampleMatch(id: string, size: number): DriveListMatch {
  return { id, name: "spec.pdf", mimeType: "application/pdf", size };
}

describe("ensureAppFolder", () => {
  it("creates the folder when none exists, then caches the id", async () => {
    const http = makeHttp();
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    const id1 = await adapter.ensureAppFolder();
    const id2 = await adapter.ensureAppFolder();
    expect(id1).toBe("folder-xyz");
    expect(id2).toBe("folder-xyz");
    expect(http.findAppFolder).toHaveBeenCalledTimes(1);
    expect(http.createFolder).toHaveBeenCalledTimes(1);
  });

  it("returns the pre-existing folder id without creating a duplicate", async () => {
    const http = makeHttp({ findAppFolder: vi.fn(async () => "folder-existing") });
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    expect(await adapter.ensureAppFolder()).toBe("folder-existing");
    expect(http.createFolder).not.toHaveBeenCalled();
  });

  it("rejects when no token is available", async () => {
    const adapter = new GoogleDriveAdapter({
      http: makeHttp(),
      token: async () => null,
      backoff: ZERO_DELAY,
    });
    await expect(adapter.ensureAppFolder()).rejects.toBeInstanceOf(StorageException);
  });
});

describe("upload — small file path", () => {
  let adapter: GoogleDriveAdapter;
  let http: DriveHttpClient;

  beforeEach(() => {
    http = makeHttp();
    adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
  });

  it("uses simpleUpload for files under the threshold", async () => {
    const file = await adapter.upload({
      bytes: new Uint8Array([1, 2, 3, 4]),
      name: "small.pdf",
      mimeType: "application/pdf",
      sizeBytes: 4,
    });
    expect(http.simpleUpload).toHaveBeenCalledTimes(1);
    expect(http.startResumableUpload).not.toHaveBeenCalled();
    expect(file.externalId).toBe("file-1");
  });

  it("refuses simple upload when only a stream is provided", async () => {
    await expect(
      adapter.upload({
        stream: async function* () {
          yield new Uint8Array(4);
        },
        name: "small.pdf",
        mimeType: "application/pdf",
        sizeBytes: 4,
      }),
    ).rejects.toThrow(/file\.bytes/);
  });
});

describe("upload — resumable path", () => {
  it("PUTs every chunk and returns the final metadata", async () => {
    const http = makeHttp({
      // First two chunks return null; the last one returns metadata.
      putChunk: vi.fn(async (uri, chunk, start, total) => {
        const end = start + chunk.byteLength;
        if (end === total) return sampleMatch("file-resumable", total);
        return null;
      }),
    });
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    const oneChunk = new Uint8Array(SIMPLE_UPLOAD_THRESHOLD_BYTES + 16);
    const file = await adapter.upload({
      stream: async function* () {
        // Single yield equal in size to the file — the adapter splits
        // internally into 8 MB chunks.
        yield oneChunk;
      },
      name: "big.bin",
      mimeType: "application/octet-stream",
      sizeBytes: oneChunk.byteLength,
    });
    expect(http.startResumableUpload).toHaveBeenCalledTimes(1);
    expect(http.putChunk).toHaveBeenCalled();
    expect(file.sizeBytes).toBe(oneChunk.byteLength);
  });

  it("throws when the stream delivers fewer bytes than advertised", async () => {
    const http = makeHttp({
      putChunk: vi.fn(async () => null), // never returns final metadata
    });
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    const promise = adapter.upload({
      stream: async function* () {
        yield new Uint8Array(SIMPLE_UPLOAD_THRESHOLD_BYTES + 1);
      },
      name: "short.bin",
      mimeType: "application/octet-stream",
      sizeBytes: SIMPLE_UPLOAD_THRESHOLD_BYTES + 64, // claims more than streamed
    });
    await expect(promise).rejects.toBeInstanceOf(StorageException);
  });
});

describe("getDownloadUrl", () => {
  it("embeds the access token and returns a Drive media URL", async () => {
    const http = makeHttp();
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    const url = await adapter.getDownloadUrl("file-abc");
    expect(url).toContain("file-abc");
    expect(url).toContain("access_token=ya29.token");
    expect(http.getMetadata).toHaveBeenCalledTimes(1);
  });

  it("rejects ttl outside (0, 3600]", async () => {
    const adapter = new GoogleDriveAdapter({
      http: makeHttp(),
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    await expect(adapter.getDownloadUrl("file-abc", 0)).rejects.toBeInstanceOf(StorageException);
    await expect(adapter.getDownloadUrl("file-abc", 4000)).rejects.toBeInstanceOf(StorageException);
  });

  it("rejects empty externalId", async () => {
    const adapter = new GoogleDriveAdapter({
      http: makeHttp(),
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    await expect(adapter.getDownloadUrl("  ")).rejects.toBeInstanceOf(StorageException);
  });
});

describe("retry on 5xx / 429", () => {
  it("retries simple upload until success", async () => {
    let attempts = 0;
    const http = makeHttp({
      simpleUpload: vi.fn(async () => {
        attempts += 1;
        if (attempts < 3) {
          throw new StorageException(StorageError.upstream(503, "Service Unavailable"));
        }
        return sampleMatch("file-retry", 4);
      }),
    });
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    const file = await adapter.upload({
      bytes: new Uint8Array(4),
      name: "f.bin",
      mimeType: "application/octet-stream",
      sizeBytes: 4,
    });
    expect(file.externalId).toBe("file-retry");
    expect(attempts).toBe(3);
  });

  it("does NOT retry on 401 — surfaces Unauthorized immediately", async () => {
    let attempts = 0;
    const http = makeHttp({
      simpleUpload: vi.fn(async () => {
        attempts += 1;
        throw new StorageException(StorageError.unauthorized("Token expired"));
      }),
    });
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    await expect(
      adapter.upload({
        bytes: new Uint8Array(4),
        name: "f.bin",
        mimeType: "application/octet-stream",
        sizeBytes: 4,
      }),
    ).rejects.toBeInstanceOf(StorageException);
    expect(attempts).toBe(1);
  });

  it("gives up after maxAttempts", async () => {
    let attempts = 0;
    const http = makeHttp({
      simpleUpload: vi.fn(async () => {
        attempts += 1;
        throw new StorageException(StorageError.upstream(503, "Service Unavailable"));
      }),
    });
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    await expect(
      adapter.upload({
        bytes: new Uint8Array(4),
        name: "f.bin",
        mimeType: "application/octet-stream",
        sizeBytes: 4,
      }),
    ).rejects.toBeInstanceOf(StorageException);
    // Initial call + 3 retries = 4 attempts in total
    expect(attempts).toBe(4);
  });

  it("retries 429 (rate limited)", async () => {
    let attempts = 0;
    const http = makeHttp({
      simpleUpload: vi.fn(async () => {
        attempts += 1;
        if (attempts === 1) throw new StorageException(StorageError.upstream(429, "Slow down"));
        return sampleMatch("file-1", 4);
      }),
    });
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    await adapter.upload({
      bytes: new Uint8Array(4),
      name: "f.bin",
      mimeType: "application/octet-stream",
      sizeBytes: 4,
    });
    expect(attempts).toBe(2);
  });
});

describe("delete", () => {
  it("forwards the externalId to the wire", async () => {
    const http = makeHttp();
    const adapter = new GoogleDriveAdapter({
      http,
      token: async () => "ya29.token",
      backoff: ZERO_DELAY,
    });
    await adapter.delete("file-abc");
    expect(http.delete).toHaveBeenCalledWith("ya29.token", "file-abc");
  });
});
